Approved/implicitly-scoped-qubit-allocation.md

```diff
@@ -0,0 +1,342 @@
+
+```qsharp
+operation UsingAndBorrowing() : Result {
+    use a = Qubit();
+    borrow b = Qubit();
+    // ...
+    return M(a); // a and b are released here.
+}
+```
+
+Example 5b: A qubit allocated in new block and released before the end of the outer block.
+
+```qsharp
+operation DifferentLifetime() : Result {
+    mutable r = Zero;
+    {
+        use q = Qubit();
+        // Apply operations...
+        set r = M(q);
+        // q is released here.
+    }
+    // Classical post-processing...
+    return r;
+}
+```
+
+# Implementation
+
+The proposal could be implemented in one of two ways:
+
+1. As syntactic sugar for the existing block statements.
+   The new `use` and `borrow` statements are transformed in the AST into `using` and `borrowing` block statements, taking in the remaining statements of the block automatically.
+   The scope statement could be transformed into `if (true) { }`.
+   This would require no change to code generation.
+2. As new AST nodes, similar to `let` and `mutable` nodes for classical values.
+   A new scope statement AST node is added.
+   Code generation must release qubits at the end of the scope, but this is already analogous to reference counting for certain data types like arrays in QIR.
+
+Both options seem similar in terms of complexity and implementation timeline, but option 2 seems like the cleaner solution.
+Neither option has any performance differences compared to the existing implementation.
+
+## Timeline
+
+There are no dependencies on other proposals or libraries.
+The work needed for either option appears to be minimal, since it is similar to the existing implementation.
+
+# Further Considerations
+
+## Related Mechanisms
+
+This proposal relies on the existing qubit management functionality that is already a core feature of Q#.
+It does not change the behavior of qubit management; it only provides new syntax for existing functionality.
+
+## Impact on Existing Mechanisms
+
+The scope statement added in this proposal can also be used to limit the scope of non-qubit variable bindings declared with `let` and `mutable`.
+This may be useful for limiting the lifetime of large data structures like arrays or for otherwise communicating the intent that a variable should only be used in a specific section within a callable.
+
+## Anticipated Interactions with Future Modifications
+
+Because the proposed modification is similar to existing functionality, it is not expected to have any impact on future language development.
+
+## Alternatives
+
+### Alternative 1: Add new statements while preserving existing block syntax
+
+Q# could add non-block `using` and `borrowing` statements while preserving existing syntax:
+
+```qsharp
+using q = Qubit();
+// ...
+
+using (q = Qubit()) {
+    // ...
+}
+
+borrowing q = Qubit();
+// ...
+
+borrowing (q = Qubit()) {
+    // ...
+}
+```
+
```

Comment:
Perhaps an odd idea on how to adapt Alternative 1 to address the issue raised at L324 (apologies in advance if it's *too* odd), but could the following two examples be equivalent?

```
// ...
{
    use q = Qubit();
    // ...
}
// ...
```

```
// ...
use q = Qubit() in {
    // ...
}
// ..
```

That is, would it be possible to make the distinction between a statement and a block the use of the `in` keyword to explicitly denote a scope? Doing so would possibly also allow for more functional-style `let` blocks, perhaps?

```
let foo = bar in {
    // foo has the value bar here.
}
// foo is unbound here.
```

Reply:
F# used to work this way; let statements explicitly introduced a new block. They made this optional in 2.0, I think, and then dropped it entirely later on. I think the "bound until end of the containing scope" style is much easier to write and read, partly because it leads to less indentation -- less visual noise.

Reply:
F# still works this way, actually (but it is optional):

```fsharp
let x = 10 in
    printfn "x is %i" x
```

To be more precise, the full form is `let x = v in e`, where `e` is an expression (not a block). My understanding is that

```fsharp
let x = 10
printfn "x is %i" x
```

is actually just syntactic sugar for the first version. The first version is still useful sometimes when you want to define a variable and use it on the same line, for example:

```fsharp
[ 1; 2; 3 ]
|> List.map (fun x -> let y = 2 * x in y * y)
```

Reply:
And playing around with this a little more, I realized that

```fsharp
let x = 10 in
    printfn "x is %i" x
printfn "x is %i" x
```

actually prints "x is 10" twice. I think this is because the `in` absorbs all the remaining expressions even if they don't match the initial indentation. This might be to support OCaml-style nested let bindings:

```fsharp
let x = 10 in
let y = x + 1 in
printfn "%i %i" x y
```

So it's not exactly the same as @cgranade's suggestion, but adding a `do` makes it indentation-sensitive, and `x` is undefined on the last line:

```fsharp
do let x = 10 in
    printfn "x is %i" x
printfn "x is %i" x
```

Reply:
> F# still sort of works this way (but it is optional):

That's kind of what I was going towards. Having something that works well for low visual noise like statement-level declarations is a great idea, but sometimes (like in the phase kickback example) it can be really nice to be very explicit about the scope of a qubit in a visually obvious way.

```Q#
use flag = Qubit() in {
    within {
        H(flag);
        X(flag);
    } apply {
        oracle();
    }
}
```

Reply:
I think my preference is to compose `{ }` and `use` as separate statements that can be used together or not, rather than having two variants of `use` (one ending with `in { }` and one ending with `;`). I do prefer `let ... in ...` as an expression that takes a binding and another expression, but it feels weird to me as a statement that takes a binding and a block. The `in` version makes really makes me wish I could do

```qsharp
Foo(let x = 10 in x * x);
```

but that wouldn't be allowed in the statement version.

Reply:
At this point, the block form of `use` is pretty much the same as the current `using` statement. Maybe we should we just keep `using` for this form and reserve `use` for the statement form?

Reply:
@samarsha: Agreed, the opening brace is definitely clear enough to denote a new scope. I like that the new form also requires one less pair of parentheses, making it a lot cleaner to read. As for `use` versus `using`, I think `use` and `borrow` have a nice parity with `let` and `set`, such that my suggestion would be to adopt `use` and `borrow`. For backwards compatibility, we could maybe keep `using` and `borrowing` as deprecated keywords?

Reply:
Could someone please help me understand why do we need to keep the block instead of the original proposal?
```qsharp
use q = Qubit();
```

Reply:
@viktorveis: Both forms are quite useful in their own respects, I think. For the case in which a qubit allocation scope is the same as an operation scope, the statement form eliminates a lot of unneeded levels of nesting. That said, whenever you're doing something like phase kickback, where a temporary qubit is only needed for a short time, retaining a block form of `use` and `borrow` can help make sure that the temporary qubit is released as soon as possible and can make visually clear the scope in which the temporary qubit is used.

Reply:
> Maybe we should we just keep `using` for this form and reserve `use` for the statement form?

I would like to be more economical in keyword usage than this, since I think needing to remember whether to write `use` or `using`, or `borrow` or `borrowing`, would be an unnecessary burden on users, when the compiler is capable of figuring out whether you meant the block or non-block form based only on whether a block follows the statement or not.

While Q# has introduced very similar keywords in the past - like `Controlled` vs `Ctl` vs `controlled`, I am not so sure that's a good idea, because personally I find having to remember which spelling of "controlled" I need in this particular context is kind of annoying.

Reply:
Does having this syntax:
```
use q = Qubit() {
    // ...
}
```
mean that we will support the following syntax in the future?
```
let a, b, c = 12, "Yes", 3.0 {
    // ...
}
```
where `a`, `b`, and `c` are scoped to to following block? Seems like we already have a syntax similar to that with the specialization declaration syntax:
```
operation Foo(a : Int, b : String, c : Double) : Unit is Ctl {
    body(...) {
        // The contents of the parentheses define all the bindings at the start of this scope.
        // The "..." bring in all the super-scope's bindings.
    }
    controlled (control, ...) {
        // Here, in addition to the super-scope's bindings, an additional binding for the control qubits is included.
    }
}

Reply:
If this could be adapted to a more general form that would allow assignment, it could be used in general to define callable definitions, specialization definitions, qubit lifetimes, and any arbitrary scopes that a coder may want to initialize, possibly for the purpose of shadowing variables. It could really simplify the construction of the language. Is this of interest?

Reply:
> Does having this syntax:
> 
> ```
> use q = Qubit() {
>     // ...
> }
> ```
> 
> mean that we will support the following syntax in the future?
> 
> ```
> let a, b, c = 12, "Yes", 3.0 {
>     // ...
> }
> ```

There would be consistency from that, but I don't think having a block `use` statement necessarily means we will or should support a block `let` statement.

> If this could be adapted to a more general form that would allow assignment, it could be used in general to define callable definitions, specialization definitions, qubit lifetimes, and any arbitrary scopes that a coder may want to initialize, possibly for the purpose of shadowing variables.

I am not sure what the more general form you're referring to is, can you clarify?

Reply:
I wasn't referring to something specific, just an idea to consider, if we wanted to go down that route. However the scope syntax being suggested in this PR is pretty general and could be used in a variety of places, so it may cover what I'm suggesting here. I like the idea of having the generic scope syntax being suggested over the `use` blocks as I think it will provide more consistency across the language's syntax.

Reply:
Having come this far -- which I think is quite reasonable -- is there a reason we don't continue with `using` for the block case and `use` for the statement case? This would reduce the amount of churn in user and library code at the cost of two additional keywords.
For what it's worth, there's a reasonable "English" reason for `using` being present progressive -- something that's true during a process -- and `use` being a command, "do it now"...
If we ever want block-form `let`, I think this would argue for `letting` being the keyword, which actually doesn't feel wrong to me. `Mutating` is a little bit more odd -- although it's not clear to me how much value there is in such a construct in actual use.

All that said, I'm also fine with having `use` be the common keyword between the two forms.

Reply:
@alan-geller Possibly my reply was lost since this is a long thread, but here was my reply to having both `use` and `using`:

> > Maybe we should we just keep `using` for this form and reserve `use` for the statement form?
> 
> I would like to be more economical in keyword usage than this, since I think needing to remember whether to write `use` or `using`, or `borrow` or `borrowing`, would be an unnecessary burden on users, when the compiler is capable of figuring out whether you meant the block or non-block form based only on whether a block follows the statement or not.
> 
> While Q# has introduced very similar keywords in the past - like `Controlled` vs `Ctl` vs `controlled`, I am not so sure that's a good idea, because personally I find that having to remember which spelling of "controlled" I need in a particular context is confusing.

Reply:
Regarding churn in existing Q# code, my primary concern while we are pre-1.0 is to make sure we have a cohesive design (e.g. without redundant keywords). I view backwards compatibility as a secondary concern to that. Once we reach 1.0, and no longer have the freedom to make minor adjustments like this, those two priorities will switch places IMO.